doc/usage.md

```diff
@@ -771,7 +792,10 @@ SIGHUP
 :   Reload config. Same as issuing the command **RELOAD** on the console.
 
 SIGINT
-:   Safe shutdown. Same as issuing **PAUSE** and **SHUTDOWN** on the console.
+:   Safe shutdown. If `so_reuseport` is set to 0, then this is the same as
+    issuing **PAUSE** and **SHUTDOWN** on the console. If `so_reuseport` is set
+    to 1 then this is the same as issuing **SHUTDOWN WAIT_FOR_CLIENTS** on
+    the console.
 
```

Comment:
Hmm, we should use a separate signal for this. Otherwise, it's too much magic.

I think the best way would be:
- Keep SIGINT as is.
- Move the current SIGTERM behavior to SIGQUIT.
- This new behavior goes to SIGTERM.

It's a compatibility break, but it has some advantages. It restores the intended progression that SIGTERM is "softer" than SIGINT. And it creates some consistency with the postgres/pg_ctl shutdown modes.

We could even have a setting like signal_compat that keeps the old signal assignments. Not sure if it's needed, but it would be possible pretty easily.

Reply:
While I understand the desire to be in line with how Postgres uses signals, I don't think breaking compatibility in such a big way is worth it. Changing SIGTERM would start making a lot of Docker and SystemD stop commands suddenly change their behaviour. I don't think that's something we should be doing.

The reason I suggested changing SIGINT, is because if `so_reuseport` because there are multiple processes listening is that the current SIGINT behaviour is actively harmful. Because the PgBouncer process is still accepting new connections only to then send an error on them, while one of the other listening procesess might still be accepting connections.

Reply:
I just checked:
- https://hub.docker.com/r/bitnami/pgbouncer/ uses SIGTERM (afaict)
- pgdg apt systemd service file uses SIGINT
- pgdg rpm systemd service file uses SIGINT

So changing behaviour of any of them seems pretty likely to cause problems in production for some users. I think assigning the new behaviour to SIGQUIT is the least disruptive.

It would be possible to add some config compat flag, to change the meaning of the signals to match postgres its intent. I think that's only likely to cause more confusion, because now you have to know what the config is to know how to stop pgbouncer in the way that you want.

Reply:
> I just checked:
> 
> * https://hub.docker.com/r/bitnami/pgbouncer/ uses SIGTERM (afaict)

I didn't find the source code for this, but if it's true, then I think it's possible that it's a mistake, possibly exactly because pgbouncer does SIGTERM and SIGINT backwards.

> * pgdg apt systemd service file uses SIGINT
> * pgdg rpm systemd service file uses SIGINT
> 
> So changing behaviour of any of them seems pretty likely to cause problems in production for some users.

My proposal would leave SIGINT unchanged, and almost everybody should be using SIGINT normally.

I'm not saying I'm 100% comfortable, but it doesn't seem too bad.

Reply:
Okay I implemented this proposal now, and if we put a warning in the release notes I think we should be fine.

The only issue I have is that Windows does not have SIGQUIT, so now you cannot do an immediate shutdown in windows anymore. Which maybe is fine.

Reply:
Okay, I think I found (and implemented) a rather elegant solution to the windows problem: A second SIGTERM signal will trigger an immediate shutdown.

Reply:
Hmm, turns out windows doesn't actually send SIGTERM requests. So this was a non-issue. Still, it seems quite nice to have this "immediate" shutdown happen on repeated signals. I think I'd actually go so far as also wanting to implement this for SIGINT. Having double Ctrl+C do an immediate shutdown seems really nice for interactive usages.

Reply:
I implement "double SIGINT" == "immediate shutdown"